sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/LargeBlobTest.groovy

```diff
@@ -232,20 +320,57 @@ class LargeBlobTest extends APISpec {
         Mono<HttpResponse> process(HttpPipelineCallContext httpPipelineCallContext, HttpPipelineNextPolicy httpPipelineNextPolicy) {
             def request = httpPipelineCallContext.httpRequest
             // Substitute large body for put block requests and collect size of original body
-            if (request.url.getQuery() != null && request.url.getQuery().endsWith("comp=block")) {
-                if (collectSize) {
-                    def bytesReceived = request.getBody().reduce(0L, new BiFunction<Long, ByteBuffer, Long>() {
-                        @Override
-                        Long apply(Long a, ByteBuffer byteBuffer) {
-                            return a + byteBuffer.remaining()
-                        }
-                    })
-                    putBlockPayloadSizes.add(bytesReceived)
+            def urlString = request.getUrl().toString()
+            if (isPutBlockRequest(request)) {
+                if (!retryTracker.get(urlString, false)) {
+                    blocksCount.incrementAndGet()
+                    retryTracker.put(urlString, true)
+                }
+                Mono<Long> count = interceptBody(request)
+                if (count != null) {
+                    return count.flatMap { bytes ->
+                        putBlockPayloadSizes.add(bytes)
+                        return httpPipelineNextPolicy.process()
+                    }
+                }
+            } else if (isSinglePutBlobRequest(request)) {
+                if (!retryTracker.get(urlString, false)) {
+                    singleUploadCount.incrementAndGet()
+                    retryTracker.put(urlString, true)
+                }
+                Mono<Long> count = interceptBody(request)
+                if (count != null) {
+                    return count.flatMap { bytes ->
+                        putBlobPayloadSizes.add(bytes)
+                        return httpPipelineNextPolicy.process()
+                    }
                 }
-                count.incrementAndGet()
-                request.setBody("dummyBody")
             }
             return httpPipelineNextPolicy.process()
         }
+
+        private Mono<Long> interceptBody(HttpRequest request) {
+            Mono<Long> result = null
+            if (collectSize) {
+                result = request.getBody().reduce(0L, new BiFunction<Long, ByteBuffer, Long>() {
+                    @Override
+                    Long apply(Long a, ByteBuffer byteBuffer) {
+                        return a + byteBuffer.remaining()
+                    }
+                })
+            }
+            request.setBody("dummyBody")
+            return result
+        }
+
+        private boolean isPutBlockRequest(HttpRequest request) {
+            return request.url.getQuery() != null && request.url.getQuery().endsWith("comp=block");
```

Comment:
Should this check that the query contains `comp=block`.

Reply:
Make sense. Thank you!